src/mongocxx/test/client_side_encryption.cpp

```diff
@@ -322,28 +360,569 @@ TEST_CASE("Datakey and double encryption", "[client_side_encryption]") {
+    coll.insert_one(under_16mib_doc.view());
+
+    // Insert the document limits/limits-doc.json concatenated with
+    // { "_id": "encryption_exceeds_16mib",
+    //   "unencrypted": < the string 'a' repeated (16777216 - 2000) times > }
+    auto encryption_exceeds_16mib = document{} << "_id"
+                                               << "encryption_exceeds_16mib"
+                                               << "unencrypted" << under_16mib
+                                               << concatenate(limits_doc.view()) << finalize;
+
+    // Expect this to fail since encryption results in a document exceeding
+    // the maxBsonObjectSize limit.
+    REQUIRE_THROWS(coll.insert_one(encryption_exceeds_16mib.view()));
 }
 
 TEST_CASE("Views are prohibited", "[client_side_encryption]") {
-    // TODO CXX-2021
+    instance::current();
+
+    // Create a MongoClient without encryption enabled (referred to as client).
+    class client client {
+        uri {}
+    };
+
+    if (test_util::get_max_wire_version(client) < 8) {
+        // Automatic encryption requires wire version 8.
+        WARN("Skipping - max wire version is < 8");
+        return;
+    }
+
+    // Using client, drop and create a view named db.view with an empty pipeline.
+    // E.g. using the command { "create": "view", "viewOn": "coll" }.
+    auto db = client["db"];
+    db["view"].drop();
+
+    auto cmd = make_document(kvp("create", "view"), kvp("viewOn", "coll"));
+    db.run_command(cmd.view());
+
+    // Create a MongoClient configured with auto encryption (referred to as client_encrypted)
+    // Configure with the local KMS provider as follows:
+    // { "local": { "key": <base64 decoding of LOCAL_MASTERKEY> } }
+    // Configure with the keyVaultNamespace set to keyvault.datakeys.
+    options::client opts;
+    auto kms_doc = _make_kms_doc();
+    _add_client_encrypted_opts(&opts, {}, std::move(kms_doc));
+    class client client_encrypted {
+        uri{}, std::move(opts)
+    };
+
+    // Using client_encrypted, attempt to insert a document into db.view.
+    // Expect an exception to be thrown containing the message: "cannot auto encrypt a view".
+    REQUIRE_THROWS_MATCHES(client_encrypted["db"]["view"].insert_one(cmd.view()),
+                           mongocxx::exception,
+                           test_util::mongocxx_exception_matcher{"cannot auto encrypt a view"});
 }
 
 TEST_CASE("Corpus", "[client_side_encryption]") {
-    // TODO CXX-2021
+    // TODO CXX-2022
+}
+
+void _round_trip(class client_encryption* client_encryption,
+                 bsoncxx::types::bson_value::view datakey) {
+    auto to_encrypt_doc = document{} << "v"
+                                     << "test" << finalize;
+    auto to_encrypt = to_encrypt_doc.view()["v"].get_value();
+
+    options::encrypt encrypt_opts;
+    encrypt_opts.algorithm(options::encrypt::encryption_algorithm::k_deterministic);
+    encrypt_opts.key_id(datakey.get_binary());
```

Comment:
Actually, related to my comment about ownership above, IIUC this could also be problematic. The underlying bytes that b_binary points to are unowned. I imagine this is also UB:

```
{
   bsoncxx::types::bson_value::value data_key = client_encryption->create_data_key("aws", std::move(opts));
   encrypt_opts.key_id (data_key.get_binary());
}
// encrypt_opts is holding onto binary data that has been destroyed.
```

I think `encrypt_opts` would need to make a copy of the underlying data.

This seems out of scope of this ticket. But if you agree that this seems like a bug, I think we should address this before releasing. Perhaps a straightforward solution is to have the argument of `key_id` be a `bsoncxx::types::bson_value::view_or_value`? It also may be an API improvement in that you can directly pass the return value of `create_data_key` as the option for the key id.

Reply:
Ah good catch!  This is definitely a concern.  I agree, we should change this method to take a view_or_value.  I opened [CXX-2063](https://jira.mongodb.org/browse/CXX-2063) to do this work before the release.

Reply:
Thanks!